compiler/AST/ForallStmt.cpp

```diff
@@ -286,6 +291,22 @@ ForallStmt* isForallIterVarDef(Expr* expr) {
 
 // Return a ForallStmt* if 'expr' is its iterable-expression.
 ForallStmt* isForallIterExpr(Expr* expr) {
+  if (CallExpr *iterCall = toCallExpr(expr)) {
+    if (iterCall->isPrimitive(PRIM_ZIP)) {
+      // in many cases PRIM_ZIP is right before the pertinent forall stmt.
+      // However, when the forall has a reduce intent, its setup comes in
+      // between the PRIM_ZIP and the forall statement.
+      Expr *cur = expr->next;
+      while(cur) {
+        if (ForallStmt* pfs = toForallStmt(cur)) {
+          INT_ASSERT(pfs->zipCall == iterCall);
+          return pfs;
```

Comment:
The above assert is asking for trouble. I suggest to change it to a conditional, i.e.
`if pfs->zipCall == iterCall then return pfs;`

Please DO think about the situation where there are two ForallStmts in the same block.
Are you guaranteed that one of them will not occur between the other ForallStmt and its zipCall ?
If you believe so, then add `else return nullptr;` to the above conditional.

Reply:
Thinking some more about it, the above code should not even be there, assuming that the ForallStmt owns its zipCall. Instead it should just do:
```c++
if (ForallStmt* pfs = toForallStmt(expr->parentExpr)
  if (expr == pfs->zipCall)
    return pfs;
```

Then, the condition `if (ForallStmt* pfs = toForallStmt(expr->parentExpr))`
can be shared with the other check below.

Reply:
I believe github is playing tricks on us. The code that is shown above your comment is not even there and something that I have changed to be something similar to what you have suggested in your second comment already.

Reply:
This is the commit that has changed this part, and when I look at it, your comment looks at the removed part of the code. https://github.com/chapel-lang/chapel/pull/17212/commits/a6ea9bb91694bd3d5922efbda0873285f69fe339

Reply:
You are right.

Among all the tests, I would check `if (ForallStmt* pfs = toForallStmt(expr->parentExpr))` first as it will filter out most of the irrelevant exprs. However any order of the checks should work.

Reply:
Ah good point about the order of checks. I'll shuffle them the way you suggest.